### PR TITLE
Added CopyOnSelect as a Global Setting

### LINKED ISSUE #524: Feature: add copy on select (per Unix terminals)
Similar to #523, except the reverse: selected data in the terminal is not copyable to other apps.

## Your Windows build number:  (Type `ver` at a Windows Command Prompt)

Can't tell you, because when I select it it won't paste into this form!

But using another terminal app (and fixing the command to work with Windows default shell):

```
$ #[Environment]::OSVersion.VersionString
Microsoft Windows NT 10.0.18890.0
```

## What you're doing and what's happening: (Copy & paste specific commands and their output, or include screen shots)

Both:

 - Highlighting text to select it then pasting it fails
 - Highlighting text, pressing Ctrl C (which shouldn't be necessary) and then pasting it fails 

## What's wrong / what should be happening instead:

Different data is pasted

It should be pasting whatever was selected in the terminal.

Here's the options provided by Fluent Terminal:

Annotation 2019-07-21 175305

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
Introduces the "copyOnSelect" setting to "globals" as a boolean type. When set to `true`, a selection is immediately copied to your clipboard upon creation. When set to `false`, the selection persists and awaits further action.

## PR Checklist
* [x] Closes #524 
* [x] CLA signed
* [x] Tests ~added/passed~ not necessary
* [x] Requires documentation to be updated
* [x] I am a core contributor

## Detailed Description of the Pull Request / Additional comments
Standard "added a new global setting" changes
`TermControl`'s `PointerReleased` event is updated to support a mouse button release event. Just hooked up the copy function right there.

## Validation Steps Performed
- `copyOnSelect` = true
  - make selection
  - right click to paste (selection persists)
  - Works! ✔
- `copyOnSelect` = true
  - make selection
  - right click to paste
  - left click on a cell
  - right click to paste
  - same data should be pasted twice
  - Works! ✔
- `copyOnSelect` = true
  - click on a cell (no selection)
  - drag to another cell (selection visible)
  - drag back to first cell (selection should persist, but should only cover one cell)
  - Works! ✔
- `copyOnSelect` = false
  - make selection
  - selection persists
  - right click to copy (selection disappears)
  - right click to paste
  - Works! ✔

### CODE REVIEW

**Comment by mcpiroman** [comment]
Should the selection disappear? That's not how it works on other terminals.

**Comment by carlos-zamora** [comment]
> Should the selection disappear? That's not how it works on other terminals.

Yeah. I originally had it disappear to provide feedback that you copied it. I'm going to change it to:
- selection persists after copy
- right-click immediately pastes (doesn't do a redundant copy)
- click/tap (no drag) should...
  - not create a new selection
  - @DHowett-MSFT should it erase old selection?

**Comment by carlos-zamora** [comment]
Question: should setting be renamed to `copyMode` = {`Windows`, `Unix`}? 👍 or 👎 and put your thoughts below!

**Comment by zadjii-msft** [comment]
I think I personally prefer `copyOnSelect`. It's much more literally what I want - I wouldn't necessarily know that `unix` means "copy on select". Is there other possible values of `copyMode` you're thinking of other than Windows vs Unix?

**Comment by carlos-zamora** [comment]
> I think I personally prefer `copyOnSelect`. It's much more literally what I want - I wouldn't necessarily know that `unix` means "copy on select". Is there other possible values of `copyMode` you're thinking of other than Windows vs Unix?

IIRC, MacOS's Terminal has a drop-down on right-click. But that's also weird because there'd be different functionality through that drop down menu like "search with Google", I think. I don't remember what the other drop-down options are.

**Comment by 0xabu** [comment]
> Question: should setting be renamed to copyMode = {Windows, Unix}?

A copyMode of "unix" is pretty ambiguous, IMO it could just as easily refer to line ending conversion as it could to the select/click behaviour. Also, AFAIK (grumpy old man tone) copy-on-select derives more from X and its ancestors than it does Unix.

**Comment by carlos-zamora** [comment]
Alright, decided on the following model for the settings:
- `copyOnSelect: bool` --> when enabled, this makes it so that...
  - when you make a selection, it gets copied
  - a single cell selection is not supported (unless you select more than a cell then drag the mouse back to get a single cell)
  - right-click always pastes
- `pauseOnSelect: bool` (#980) --> when enabled, this makes it so that...
  - output is paused when a selection is displayed/active
- pointer bindings (#1553) --> we'll have a bunch of actions supported. A few include...
  - right-click (as an example) being bound to...
    - `paste`: always paste
    - `smartCopyPaste`: copy, if selection active. Paste otherwise.
    - `menu`: context menu as shown above

The only tricky part I see here is `copyOnSelect`'s 3rd function potentially interfering with pointer bindings. For now, I say we keep it as is. When pointer bindings comes along, we can make it so that `smartCopyPaste` specifically clears the selection. Thus kinda breaking this up into 2 things (not a big fan but I feel that that's the best way to approach this issue).

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 1]
```diff
@@ -121,6 +128,24 @@ const SHORT Terminal::_ExpandWideGlyphSelectionRight(const SHORT xPos, const SHO
...
+const bool Terminal::IsSingleCellSelection() const noexcept
+{
+    return (_selectionAnchor == _endSelectionPosition);
+}
+
+// Method Description:
+// - Checks if selection on a single cell is allowed for rendering purposes
+// Return Value:
+// - bool representing if selection is only a single cell. Used for copyOnSelect
+const bool Terminal::IsSingleCellCopyAllowed() const noexcept
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalCore/TerminalSelection.cpp]: can this and the above be folded into `IsSelectionActive`?

**Comment by nguerrera** [comment]
>copyOnSelect = false
>make selection
>selection persists
>right click to copy (selection disappears)

Does pressing enter to copy instead of right click work here? I have over a decade of muscle memory pressing enter at this step, and it's currently causing me to make a lot of mistakes in the new terminal.

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 2]
```diff
@@ -160,14 +163,17 @@ class Microsoft::Terminal::Core::Terminal final :
 
     bool _snapOnInput;
 
-    // Text Selection
+#pragma region Text Selection
     COORD _selectionAnchor;
     COORD _endSelectionPosition;
     bool _boxSelection;
     bool _selectionActive;
+    bool _allowSingleCharSelection;
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalCore/Terminal.hpp]: Selection. This basically addresses the issue where people would just click on the terminal and select a single cell. We don't actually want that to be a thing in this new "copyOnSelect" mode. The only way to select a single cell is to select two cells then drag the mouse back over a single character.

**Comment by carlos-zamora** [comment]
@zadjii-msft CopyOnSelect tests got added. Note that an "active selection" is considered copy-able. Handling a single cell selection is really weird but I thing the refactor and the test should make it clear what kind of behavior is now expected when using this new mode.

**Comment by carlos-zamora** [comment]
@msftbot make sure @DHowett-MSFT  signs off

File: src/cascadia/TerminalCore/Terminal.hpp [hunk 3]
```diff
@@ -137,7 +137,10 @@ class Microsoft::Terminal::Core::Terminal final :
 
 #pragma region TextSelection
     // These methods are defined in TerminalSelection.cpp
+    const bool IsSingleCellSelection() const noexcept;
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalCore/Terminal.hpp]: should these be moved into the private compartment

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 4]
```diff
@@ -1387,24 +1398,34 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
     // - trimTrailingWhitespace: enable removing any whitespace from copied selection
     //    and get text to appear on separate lines.
     bool TermControl::CopySelectionToClipboard(bool trimTrailingWhitespace)
     {
-        if (_terminal != nullptr && _terminal->IsAreaSelected())
+        // no selection --> nothing to copy
+        // prevent single cell selection (if not allowed)
+        if (_terminal == nullptr ||
+            !_terminal->IsSelectionActive() ||
+            (!_terminal->IsSingleCellCopyAllowed() && _terminal->IsSingleCellSelection()))
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalControl/TermControl.cpp]: these two conditions are now checked by `IsSelectionActive`, right?

File: src/cascadia/TerminalCore/TerminalSelection.cpp (hunk 1 above)
Comment (DHowett-MSFT) [reply src/cascadia/TerminalCore/TerminalSelection.cpp]: this one can be deleted once the TermControl caller goes away

**Comment by timriker** [comment]
I'm running the version in the store. (0.3.2171.0) With this version, click-drag, double-click and triple-click all select and copy to clipboard. Right click removes selection, but does NOT paste. A second right click DOES paste. This is odd.

On other terminal apps on windows and Linux, middle button is paste and right button is menu. Is there an option to get that functionality?

What does middle button do currently?

I don't see a "CopyOnSelect" option in settings.

**Comment by DHowett-MSFT** [comment]
image

Our release process can't possibly be that fast.

**Comment by DHowett-MSFT** [comment]
As of 0.3.2171.0: selection just selects, right-click copies _and_ deselects, and then right-click pastes.
As of [next version]: `copyOnSelect` is added, but the preexisting behavior stays the same if it is set to `false`.

**Comment by timriker** [comment]
I didn't expect it to be in the store release yet, just being clear that I'm not yet testing the current change. :)

I would like to see more Linux-y feel with right button select modes, middle button paste mode, and right button menu. It's been requested before on other posts. I was just wondering if there's a plan for adding that. Thanks for the updates!

**Comment by DHowett-MSFT** [comment]
Thanks! We haven't yet looked at middle-paste, but we've talked about "mouse bindings" (as a post-1.0 feature)!

**Comment by alecive** [comment]
Question: I'm using 0.4.2382.0 but I am still having big issues when using `WSL` with `windows terminal`. Should I see the `copyOnSelect` toggle somewhere in the settings? Why don't I see it there?

I am just trying to fix the annoying bug of copying something in my clipboard [from e.g. the notepad or the browser] and then pasting it in my `WSL` instance within `windows terminal`. It this PR fixing this issue?

**Comment by DHowett-MSFT** [comment]
>fix the annoying bug of copying something

@alecive what you have described is not a bug (???) so we can't tell if this pull request, which _adds an option that makes it so text you select gets copied to the clipboard automatically_, fixes it.

(fwiw: add `copyOnSelect` to your profiles.json as a boolean, either in the `globals` section or not (if you don't have one))

**Comment by lagunary** [comment]
Awesome!!!!

**Pull request merged**